src/appleseed/renderer/modeling/aov/diagnosticaov.cpp

```diff
@@ -0,0 +1,339 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2018 Kevin Masson, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "diagnosticaov.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/aov/aovaccumulator.h"
+#include "renderer/kernel/aov/imagestack.h"
+#include "renderer/kernel/rendering/pixelcontext.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingresult.h"
+#include "renderer/modeling/aov/aov.h"
+#include "renderer/modeling/frame/frame.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/image/tile.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Diagnostic AOV accumulator.
+    //
+
+    class DiagnosticAOVAccumulator
+      : public AOVAccumulator
+    {
+      public:
+        DiagnosticAOVAccumulator(
+            Image& image,
+            const string& diagnostic_name)
+          : m_image(image)
+          , m_diagnostic_name(diagnostic_name)
+        {
+        }
+
+        void on_tile_end(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y)
+        {
+            for (size_t i = 0, e = frame.aov_images().size(); i < e; ++i)
+            {
+                const string aov_name = frame.aov_images().get_name(i);
+
+                if (aov_name == m_diagnostic_name)
```

Comment:
Do the linear search only the first time it is needed. The image does not change while the AOVAccumulator is alive and accumulators are not shared between threads.

const Image* m_image = nullptr;
if (!m_image) { do the search here... }